### PR TITLE
[SYCL] Remove math function decl in builtins.hpp

### PR DESCRIPTION
**Pull request by jinge90**

### CODE REVIEW

**Comment by jinge90** [comment]
Hi, @steffenlarsen
This draft PR removes all cmath decls from sycl builtins.hpp, we will give green light for these math functions in SYCL sema checking instead. Could you help review and evaluate whether we can proceed with this PR?
Thanks very much.

**Review by steffenlarsen — APPROVED** [review]
I am fine with this if the frontend people are happy with the Sema solution.

**Review by elizabethandrews — COMMENTED** [review]
Can you explain why this is required? This seems strange to me. @AaronBallman @premanandrao  can you take a look please?

**Comment by aelovikov-intel** [comment]
Putting myself in reviewers just to get the notifications matching my outlook filters, please ignore it.

**Comment by steffenlarsen** [comment]
> Can you explain why this is required? This seems strange to me.

With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with `using namespace sycl` in their code. Where `cmath` is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers.

**Comment by aelovikov-intel** [comment]
> > Can you explain why this is required? This seems strange to me.
> 
> With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with `using namespace sycl` in their code. Where `cmath` is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers.

I'm working on a different approach to this - https://github.com/intel/llvm/compare/sycl...aelovikov-intel:llvm:no-complex-include.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -492,6 +508,10 @@ static bool isSYCLUndefinedAllowed(const FunctionDecl *Callee,
   if (!Callee->getIdentifier())
     return false;
 
+  if (LibdeviceCmathSet.count(Callee->getName().str()) &&
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:511]: One thing that confuses me is: aren't all these functions already builtins? 

https://github.com/intel/llvm/blob/89e2280d53e50346e69ed8354d47a9223f46f483/clang/lib/Sema/SemaSYCL.cpp#L4680

https://github.com/intel/llvm/blob/89e2280d53e50346e69ed8354d47a9223f46f483/clang/include/clang/Basic/Builtins.def#L1350C55-L1350C55

Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:511]: Is the issue that the sycl versions of these functions don't get treated as builtins, so we need this new table to identify them?

**Comment by jinge90** [comment]
Hi, all.
We supported to use cmath/complex functions in sycl kernel code via libdevice long time. If users want to call functions from <cmath>, <complex>, they only need to use these headers and call them just like what they do in normal CPU programs.
These math functions are implemented in libm, if we directly call them in sycl kernel code, sycl semantic checking will report error saying we can't call undefined function without "sycl external" attribute, these math/complex functions' declarations are in system header which we can't modify, so we add declarations for these functions with "sycl external" attribute decorated in sycl builtins.hpp.
However, as @steffenlarsen mentioned:
"With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with using namespace sycl in their code. Where cmath is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers."
So, we are trying to remove these math decls in sycl builtins.hpp, the most straightforward approach seems to give green light to these cmath functions which has been supported by libdevice in sycl kernel.
On Linux platform, <cmath>'s implementation look like:
  inline _GLIBCXX_CONSTEXPR float
 sqrt(float __x)
 { return __builtin_sqrtf(__x); }
sema checking will not report error for such case since math function will be treated as "builtin" but MSVC's cmath implementation is different, if we directly remove these decls in builtins.hpp, cmath libdevice test suite will fail, this is why I added the table checking for math functions.
Thanks very much.

**Comment by jinge90** [comment]
> > > Can you explain why this is required? This seems strange to me.
> > 
> > 
> > With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with `using namespace sycl` in their code. Where `cmath` is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers.
> 
> I'm working on a different approach to this - [sycl...aelovikov-intel:llvm:no-complex-include](https://github.com/intel/llvm/compare/sycl...aelovikov-intel:llvm:no-complex-include).

Hi, @aelovikov-intel 
Some more explanation about intention of this PR:
We need these removed decls in builtins.hpp previously because we have to decorate them with "sycl extern", otherwise sycl sema checking will report error which will block users to call functions provided by libdevice.
Now, we remove these decls and in order to avoid triggering sycl sema checking error report, we give green light for these functions already supported by libdevice.

Thanks very much.

Thanks very much.

**Comment by elizabethandrews** [comment]
> With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with using namespace sycl in their code. Where cmath is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers.

Can you elaborate on this please? What were the recent changes which introduced this ambiguity? By conflicting builtins, do you mean the ones in sycl/builtins.hpp and libm?

Can you also answer @AaronBallman's question above? Why isn't the `!Callee->getBuiltinID()` check in SemaSYCL.cpp catching this case?

Does it make sense to avoid the SYCL_EXTERNAL error for all functions declared in system headers?

**Comment by jinge90** [comment]
> > With recent changes to SYCL builtin definitions, having conflicting math builtins in the global namespace may cause ambiguity errors for users with using namespace sycl in their code. Where cmath is included we cannot avoid this, but we should at the very least avoid declaring/including them in the SYCL headers.
> 
> Can you elaborate on this please? What were the recent changes which introduced this ambiguity? By conflicting builtins, do you mean the ones in sycl/builtins.hpp and libm?
> 
> Can you also answer @AaronBallman's question above? Why isn't the `!Callee->getBuiltinID()` check in SemaSYCL.cpp catching this case?
> 
> Does it make sense to avoid the SYCL_EXTERNAL error for all functions declared in system headers?

Hi, @elizabethandrews and @AaronBallman 
The reason why "!Callee->getBuiltinID()" didn't hit is the test added "-fno-builtin" flag.
With SYCL libdevice support, the math functions can be used in kernel code no matter whether 'fno-builtin' is added or not. We added 'fno-builtin' flag in test to make sure the behavior is unified in both case.
Currently, we can't avoid the SYCL_EXTERNAL error for all functions declared in system headers. Only a small number of functions from system header are supported in sycl kernel via libdevice, most of them are numeric related functions. For other system functions, SYCL sema should report compfail.
Thanks very much.

**Comment by AaronBallman** [comment]
> The reason why "!Callee->getBuiltinID()" didn't hit is the test added "-fno-builtin" flag.

I read this as "there's no bug to be fixed" -- if the user passes `-fno-builtin`, we should not be giving them *any* builtins because that's what they expect. What's more, because these are library builtins, the reason a user would pass `-fno-builtin` in the first place is because they're providing a symbol that would otherwise come from a builtin; so instead of calling the function the user defined, we silently call our own function that probably does mostly the same thing (so the user may not notice the issue for a long time).

**Comment by jinge90** [comment]
> > The reason why "!Callee->getBuiltinID()" didn't hit is the test added "-fno-builtin" flag.
> 
> I read this as "there's no bug to be fixed" -- if the user passes `-fno-builtin`, we should not be giving them _any_ builtins because that's what they expect. What's more, because these are library builtins, the reason a user would pass `-fno-builtin` in the first place is because they're providing a symbol that would otherwise come from a builtin; so instead of calling the function the user defined, we silently call our own function that probably does mostly the same thing (so the user may not notice the issue for a long time).

Hi, @AaronBallman 
Do you mean if users pass "-fno-builtin", then it means users will provide their own functions that would otherwise come from a builtin and if they intend to call this function in sycl kernel code, they will add "sycl external" to decorate, so sycl sema checking will not report error for this behavior and we don't need to link libdevice implementation with user's device code either.
Thanks very much!

**Pull request closed without merging**